[WIP] Integration tests for ValidatingAdmissionPolicy

DangerOnTheRanger opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR adds a suite of integration tests intended to test the implementation of `ValidatingAdmissionPolicy` (https://github.com/kubernetes/kubernetes/pull/113314)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:

This PR currently includes commits from https://github.com/kubernetes/kubernetes/pull/113314 (51f8ce8f52b4860d15fae1c97bb90afc17e47ba6 and every commit prior to that) - code pertaining to those commits should be reviewed in that PR. When that PR merges, those commits will be removed from this PR.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control
```

cici37 committed:
Copy over admissionregistration v1 to v1alpha1

cici37 committed:
Rename copy to v1alpha1

cici37 committed:
Auto updates

cici37 committed:
Adding new api version of admissionregistration.k8s.io v1alpha1 for CEL in Admission Control

cici37 committed:
Initial integration

alexzielenski committed:
Refactor cel admission plugin

cici37 committed:
Integration

DangerOnTheRanger committed:
Add initial CEL validating admission policy integration tests.

leilajal commented:
/triage accepted

andrewsykim commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,121 @@
+	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
+)
+
+func TestParamObject(t *testing.T) {
```
We should name this test to something more generic since it's not specific to using params

DangerOnTheRanger replied:
Renamed it to `TestValidation` for now, good point

andrewsykim commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	failurePolicy := admissionregistrationv1alpha1.Fail
+	policy := &admissionregistrationv1alpha1.ValidatingAdmissionPolicy{
```
We can either do this now or later when we have more cases, but create a test table here containing a policy, a policybinding, a param object and the object to create seems worthwhile doing.

DangerOnTheRanger replied:
Makes sense, if we keep this test around I definitely think it makes sense to turn it into a table test.

DangerOnTheRanger commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,121 @@
+			"namespaceSuffix": ".k8s.io",
+		},
+	}
+	if _, err := client.CoreV1().ConfigMaps("default").Create(context.TODO(), configMap, metav1.CreateOptions{}); err != nil {
```
There's `CreateNamespaceOrDie`, though I'm not sure there's a utility function that ensures namespace creation fails (definitely could have missed it while looking though).

DangerOnTheRanger commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -118,4 +118,14 @@ func TestParamObject(t *testing.T) {
+		},
+	}
+
+	if _, err := client.CoreV1().Namespaces().Create(context.TODO(), namespace, metav1.CreateOptions{}); err != nil {
```
Fixed - the test now checks to see whether namespace creation fails

jiahuif committed:
require superuser for bindings.

maxsmythe committed:
Add matching, change nil to MatchPolicy to mean 'exact', update test.

alexzielenski committed:
move lister, client injection to constructor

andrewsykim commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,274 @@
+	}
+}
+
+func TestRuntimeTypeError(t *testing.T) {
```
Curious why this test is named `TestRuntimeTypeError`

DangerOnTheRanger replied:
The test is meant to check that a CEL-side runtime error with `FailurePolicy` set to `Ignore` results in an allow. Maybe a better name for it would be `TestRuntimeExpressionError` - I've renamed it accordingly.

DangerOnTheRanger commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,274 @@
+	policy := withParams(&admissionregistrationv1alpha1.ParamKind{
+		APIVersion: "v1",
+		Kind:       "ConfigMap",
+	},
```
Whoops, fixed

DangerOnTheRanger commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,274 @@
+		},
+	}
+
+	if _, err := client.CoreV1().Namespaces().Create(context.TODO(), namespace, metav1.CreateOptions{}); err == nil {
```
Fixed, the test checks the error message to see if it was due to the rule (and not due to the rule failing at runtime either - that was happening before...)

DangerOnTheRanger commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,461 @@
+		t.Fatal(err)
+	}
+
+	/*
```
This test ideally creates a `ConfgMap` (temporarily resulting in a valid policy + policy binding) and then deletes it (invalidating the binding) - I believe that part of the controller is not implemented yet (so the code is commented out), so I'm leaving this comment here as a reminder (and in case others have further insight).

cici37 committed:
Fix err and FailurePolicy checking

jpbetz committed:
Add message support, clean up error messaging

cici37 committed:
Clean up, fix tests

alexzielenski committed:
consistent function arguments

orders arguments consitent with the Validator

cici37 committed:
Allow binding match to further filtering request

cici37 committed:
Clean up, add test

cici37 committed:
API - doc update for matchResources in ValidatingAdmissionPolicyBinding

cici37 committed:
API - Auto update

cici37 committed:
Update matching_test

jpbetz committed:
Provide versioned types to CEL for evaluation

cici37 committed:
Address comments partially

cici37 committed:
refactor admission

jpbetz committed:
Use matched GVK

jiahuif committed:
authorization for policy and binding.

alexzielenski committed:
use atomic to set the validator

it is not ideal, but it prevents a race. we will have to switch to lock-free design for graduation anyway

alexzielenski committed:
fixup tests

andrewsykim commented on test/integration/apiserver/cel/validatingadmissionpolicy_test.go:
```diff
@@ -0,0 +1,923 @@
+	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
+)
+
+func makePolicy(name string) *admissionregistrationv1alpha1.ValidatingAdmissionPolicy {
```
Can we put all these helper functions at the bottom of the file? Just makes it easier to read when the first thing in the file are the tests

cici37 committed:
Update test with admission refactor.

cici37 committed:
Fix plugin order test

cici37 committed:
Use namespacedName as key

maxsmythe committed:
Validate that rule definitions have resource rules

Also treat empty rules equivalent to nil and remove special treatment of
isBinding

Signed-off-by: Max Smythe <smythe@google.com>

cici37 committed:
Fix test on updated validation.

dims commented:
If you still need this PR then please rebase, if not, please close the PR

dims commented:
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not
